### PR TITLE
Add RequireInternal function to assert only on non AppStore builds

### PR DESCRIPTION
**Pull request by daehn**
# What's in this PR?

* Add `RequireInternal` function to assert only on non AppStore builds.
* Unfortunately we can not use the env variable `ENABLE_DEVELOPER_MENU` outside of macros, as it wouldn't be expanded on the caller side. Because of this the Swift version `requireInternal` does rely on the bundle identifier instead. It's not optimal that the two versions are based on different checks, however, the macro does not have an performance overhead in production builds (as it will be an empty implementation in that case) and it can use a `__builtin_expect` hint which makes it more performant in non production builds as well.
* Unfortunately I was unable to use `__builtin_expect` in Swift, as I got the build error that this function is neither defined for arm64 nor x86_64.

### CODE REVIEW

**Comment by daehn** [comment]
Alternatively we could define a C function with  `__attribute__((always_inline))` which would also be possible to call from Swift.

File: Source/ZMSAsserts.swift [hunk 1]
```diff
@@ -31,3 +32,28 @@ public func require(_ condition: Bool, _ message: String = "", file: String = #f
         fatal(message, file: file, line: line)
     }
 }
+
+@objc public enum Environment: UInt8 {
+    case appStore, `internal`, debug, develop, undefined
```
Comment (daehn) [inline Source/ZMSAsserts.swift]: Maybe `unknown` makes more sense than `undefined`.

**Comment by daehn** [comment]
@mikeger good, point, only checked with the framework locally. Will use the bundle id approach for both versions then.

**Comment by daehn** [comment]
@mikeger I don't think that's completely correct. The macros are defined in the header file and visible to other frameworks. The are not compiled, but instead the preprocessor will expand them on the caller site, so any environment variables used in macros will only ever be evaluated on the caller site. You can verify this by pinning the commit SHA in the UI repo, running carthage bootstrap and placing a  `RequireInternal` call somewhere.

**Comment by mikeger** [comment]
@daehn you are correct, it would work in the main app, but it would not
work in the frameworks, since for them the macros would be unwrapped on
compilation.


Sincerely,
Mihail Gerasimenko.

2017-06-28 10:13 GMT+02:00 Silvan Daehn <notifications@github.com>:

**Comment by daehn** [comment]
@mikeger yeah right, haven't thought about that.

**Comment by daehn** [comment]
@mikeger @basine updated:

* We check the bundle identifier in the macro instead of using `Environment.current.isAppStore` to ensure callers don't need to import `WireSystem` in every file this macro is used in.
* We check the appStore flag first to ensure we short circuit and don't even try to evaluate the assertion in app store builds.

**Pull request merged**